WIP #202 - Optional algorithms for <=

schillic opened the pull request:
Closes #202.

I had to find out that Julia is not very efficient for optional arguments, even if the default argument is chosen (I had assumed that this case can be statically optimized).
Hence I defined `<=` without additional arguments (the default implementation), and added a method with an additional `algorithm` argument. If you instead use
`<=(u::AbstractVector{N}, v::AbstractVector{N}; algorithm::String="quadrant")`,
you will get the following timings:
```julia
julia> using LazySets, BenchmarkTools

julia> x = rand(2); y = rand(2);

julia> @btime LazySets.le_quadrant($x, $y);
  5.173 ns (0 allocations: 0 bytes)

julia> @btime <=($x, $y); # should be equivalent to the above, but it is slower
  10.923 ns (0 allocations: 0 bytes)

julia> @btime <=($x, $y, algorithm="quadrant");
  46.827 ns (1 allocation: 96 bytes)
```

schillic commented:
We probably do not want to take the last commit. Here are the results (random numbers, hence not really comparable; but note the higher allocations):

```julia
using LazySets, BenchmarkTools
P = HPolygon();
for i in 1:100
    addconstraint!(P, LinearConstraint(rand(2), rand(1)[1]));
end
@btime σ($[-0.5, 0.5], $P);
```

`master`:   1.545 μs (25 allocations: 1.08 KiB)
this branch:   1.595 μs (39 allocations: 1.30 KiB)

mforets commented:
> I had to find out that Julia is not very efficient for optional arguments, even if the default argument is chosen (I had assumed that this case can be statically optimized).

AFAIK this is not further a performance issue in v0.7/1.0.

mforets commented:
These are the timings that i got (all on v0.7):

```julia
using LazySets, BenchmarkTools

x = rand(2); y = rand(2);

@btime LazySets.le_quadrant($x, $y);
4.172 ns (0 allocations: 0 bytes)

@btime <=($x, $y);
4.177 ns (0 allocations: 0 bytes)

@btime <=($x, $y, "quadrant");
14.584 ns (0 allocations: 0 bytes)
```

---

For `@btime σ($[-0.5, 0.5], $P);` : 

|schillic/202|master|
|-----|------|
|1.037 μs|1.225 μs|
|18 allocations: 944 bytes|18 allocations: 944 bytes|


---

Since keyword argumnts do not cause a problem (🎉) i suggest to use them instead of `le_quadrant`.
`le_quadrant.

schillic commented:
The code I added in this PR did *not* use keyword arguments, and hence it was fast.
I added a commit that uses them now, but I do not want to merge because it is still slower on v0.7:
```julia
julia> @btime LazySets.le_quadrant($x, $y);
  3.671 ns (0 allocations: 0 bytes)

julia> @btime <=($x, $y);
  8.233 ns (0 allocations: 0 bytes)

julia> @btime <=($x, $y, "quadrant");
  7.763 ns (0 allocations: 0 bytes)
```
And for the polygon constraints, in v0.7 the allocations are fine, but the runtime got worse:
```julia
# without the last commit
julia> @btime σ($[-0.5, 0.5], $P);
  753.192 ns (18 allocations: 944 bytes)

# with the last commit
julia> @btime σ($[-0.5, 0.5], $P);
  823.456 ns (18 allocations: 944 bytes)
```

mforets commented:
As per experiments, i think its a good idea to change the default for 2d vector comparison, since the "quadrant" is faster. This branch contains deprecated code though.

mforets commented:
This branch is superseded by https://github.com/JuliaReach/LazySets.jl/pull/1986. @schillic shall we close it?

schillic commented:
> This branch is superseded by #1986. @schillic shall we close it?

#1986 just removes the `atan` implementation. This branch contains new features.

mforets commented:
Should we close this as stale?
